core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -3964,6 +3965,23 @@ private void handleSuccess(PersistJob job) {
               String ufsPath = resolution.getUri().toString();
               ufs.setOwner(tempUfsPath, inode.getOwner(), inode.getGroup());
               ufs.setMode(tempUfsPath, inode.getMode());
+
+              // Check if the size is the same to guard against a race condition where the Alluxio
+              // file is mutated in between the persist command and execution
+              if (ServerConfiguration.isSet(PropertyKey.MASTER_ASYNC_PERSIST_SIZE_VALIDATION)
+                  && ServerConfiguration.getBoolean(
+                      PropertyKey.MASTER_ASYNC_PERSIST_SIZE_VALIDATION)) {
+                UfsStatus ufsStatus = ufs.getStatus(tempUfsPath);
+                if (ufsStatus.isFile()) {
+                  UfsFileStatus status = (UfsFileStatus) ufsStatus;
+                  if (status.getContentLength() != inode.getLength()) {
```

Comment:
If UFS file is correct, will it promise that the size of UFS file has the same size in Alluxio?

Reply:
Good point. This is why I added the flag. I haven't come across a case where the data size is different between UFS and Alluxio, but I could see that happening if the UFS API is providing true file size as opposed to logical file size.

Reply:
I see, in this case, do we want to set `MASTER_ASYNC_PERSIST_SIZE_VALIDATION ` to false by default?

Reply:
I'm setting it to true because I don't expect that case to come up. If it does break some special workflow, their persist jobs will fail until this is turned off.

Reply:
ok, do you mind adding some comments here in case someone met inconsistent file size here?

Reply:
Updated